tools/autograd/templates/Functions.cpp

```diff
@@ -1970,68 +1970,125 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
+    if (A.size(-1) != A.size(-2)) {
+      Tensor rhs_solve_2;
+      // We use the same trick from above for this computation
+      if (grad_Q.defined()) {
+        rhs_solve_2 = grad_Q - at::matmul(Q, Q_term);
+      } else {
+        rhs_solve_2 = -at::matmul(Q, Q_term);
+      }
+      Tensor solve_soln_2;
+      std::tie(solve_soln_2, std::ignore) = at::triangular_solve(rhs_solve_2.transpose(-2, -1), R,
                                                                /*upper=*/true, /*transpose=*/false,
                                                                /*unitriangular=*/false);
-    grad_A.add_(solve_soln_2.transpose(-2, -1));
+      grad_A.add_(solve_soln_2.transpose(-2, -1));
+    }
+    return grad_A;
+  };
+
+  //NOTE: grad of complete QR cannot exist b/c inverse not exist/unique
+  TORCH_CHECK(((r.size(-2) <= r.size(-1) && (!some)) || some),
+		     "The derivative is not implemented when nrows > ncols and complete QR. ");
+
+  auto grad_Q = grads[0];
+  auto grad_R = grads[1];
+
+  auto m = self.size(-2);
+  auto n = self.size(-1);
+
+ if (m >= n) {
+    return square_deep_case_backward(grad_Q, grad_R, self, q, r);
+  } else {
+    // For wide (m > n) input matrices A,  partition A = [X|Y] and R = [U|V]
+    // X and U are square full rank matrices. Also partition
+    // grad_R = [grad_U, grad_V] and grad_A = [grad_X | grad_Y].
+    // To obtain grad_X we reuse the gradient formula from the square case.
+    // grad_Y  Then concatenate grad_A = [grad_X, grad_Y].
+    // Formulae: grad_X = square_case_grad(grad_Q_prime, grad_U, Q, U),
+    // where grad_q_prime = grad_Q + Y @ grad_V.T
+    // and grad_Y = Q @ grad_V.
+
+    auto Y = self.narrow(-1, m, n - m);
+    auto U = r.narrow(-1, 0, m);
+    Tensor grad_Y, grad_X, grad_Q_prime;
+
+    if (grad_R.defined()) {
+	    // grad_Y starts with the value for grad_V
+	    grad_Y = grad_R.narrow(-1, m, n - m);
+	    // reuse grad_R to store grad_U
+	    grad_R = grad_R.narrow(-1, 0, m);
+	    // grad_q_prime starts with the value of Y @ grad_V.T
+	    grad_Q_prime = at::matmul(Y, grad_Y.transpose(-2, -1));
+    } else {
+	    // when grad_R is not defined then grad_V and grad_Q_prime
+	    // get initialized with zeros
+	    grad_Y = at::zeros_like(Y, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+	    grad_Q_prime = at::zeros_like(q, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+	}
+
+    if (grad_Q.defined()) {
+	    // add the grad_Q term into grad_q_prime when defined o/w is 0
+	    grad_Q_prime = grad_Q_prime + grad_Q;
+	}
+    // Calculate grad_x using the helper. Grad_r contains the grad_u value
+    grad_X = square_deep_case_backward(grad_Q_prime, grad_R, self, q, U);
+    // Calculate grad_Y. Note grad_Y contains the grad_V value
+    // And multiplication overwrites the same memory for efficiency
+    grad_Y = at::matmul(q, grad_Y);
+    // Concatenate grad_X and grad_Y to get grad_A.
+    return at::cat({grad_X, grad_Y}, -1);
   }
-  return grad_A;
```

Comment:
This line is now inside the lambda, it is the last line of the lambda, the (added) LIC 2037